catalog/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+            .unwrap_or(None)
+    }
+
+    /// Creates a new chunk with the catalog
+    ///
+    /// This function also validates 'referential integrity' - aka that the
+    /// partition
+    // referred to by this chunk exists in the catalog
+    pub fn create_chunk(&mut self, partition_key: &str, id: u32) -> Result<()> {
+        let chunks = self.chunks.get_mut(partition_key);
+
+        // TODO return a proper error here if there is no partititon
+        let chunks = chunks.expect("No such partition");
+
+        // Ensure this chunk doesn't already exist
+        if chunks.iter().any(|c| c.id() == id) {
+            panic!("chunk already exists");
+        }
+
+        chunks.push(Chunk::new(partition_key, id));
+        Ok(())
+    }
+
+    /// Removes the specified chunk from the catalog, returning the existing
+    /// chunk, if any
+    pub fn drop_chunk(&mut self, partition_key: &str, id: u32) -> Option<Chunk> {
+        let chunks = self.chunks.get_mut(partition_key);
+
+        // TODO return a proper error here if there is no partititon
+        let chunks = chunks.expect("No such partition");
+
+        let idx = chunks
+            .iter()
+            .enumerate()
+            .filter(|(_, c)| c.id() == id)
+            .map(|(i, _)| i)
+            .next();
+        if let Some(idx) = idx {
+            Some(chunks.remove(idx))
+        } else {
+            None
+        }
+    }
+
+    /// List all chunks in this database
+    pub fn chunks(&self) -> impl Iterator<Item = &Chunk> {
+        self.chunks.values().map(|chunks| chunks.iter()).flatten()
+    }
+
+    /// List all chunks in a particular partition
+    pub fn partition_chunks(&self, key: &str) -> impl Iterator<Item = &Chunk> {
+        self.chunks
+            .get(key)
+            .into_iter()
+            .map(|chunks| chunks.iter())
+            .flatten()
+    }
+
+    // List all partitions in this dataase
+    pub fn partitions(&self) -> impl Iterator<Item = &Partition> {
+        self.partitions.values()
+    }
+
+    // Get a specific partition by name
+    pub fn partition(&self, key: impl AsRef<str>) -> Option<&Partition> {
+        let key = key.as_ref();
+        self.partitions.get(key)
+    }
+
+    // Create a new partition in the catalog, returning an error if it already
+    // exists
+    pub fn create_partition(&mut self, key: impl Into<String>) -> Result<()> {
+        let key = key.into();
+
+        let entry = self.partitions.entry(key);
+        if matches!(entry, Entry::Occupied(_)) {
+            return InternalPartitionAlreadyExists {
+                partition_key: entry.key(),
+            }
+            .fail();
+        }
+
+        let chunks = self.chunks.insert(entry.key().to_string(), Vec::new());
+        //TODO proper error handling (shouldn't had a previous chunks entry)
+        assert!(chunks.is_none());
+
+        let partition = Partition::new(entry.key());
+        entry.or_insert(partition);
+
+        Ok(())
```

Comment:
I know this is a spike but  you can use the entry API like this too:

```suggestion
        let key = key.into();

        match self.partitions.entry(key.clone()) {
            Entry::Vacant(entry) => {
                let chunks = self.chunks.insert(key.clone(), Vec::new());
                //TODO proper error handling (shouldn't had a previous chunks entry)
                assert!(chunks.is_none());
                entry.insert(Partition::new(key));
            }
            Entry::Occupied(entry) => {
                return InternalPartitionAlreadyExists {
                    partition_key: entry.key(),
                }
                .fail()
            }
        }
        Ok(())
```